### PR TITLE
Add a meta testenv for metadata: style & packaging

### PR DESCRIPTION
**Pull request by lvh**
This adds some features for tox to test some meta stuff about PyOpenSSL. Specifically:

- pyflakes
- check-manifest
- pip-review, which notifies us when there are newer packages than what we install in virtualenvs
- pyroma, which should keep us generally informed about how good a packaging citizen we're being.

This isn't ready to be merged. It fails, and not as a false positive: MANIFEST.in is incomplete, and there's a few pyflakes warnings. I'm happy to do these on this PR, or on separate PRs.

### CODE REVIEW

**Comment by reaperhulk** [comment]
Conceptually we should review this and see what has already been done and what we need in 0.16. I'll add this to the milestone.

**Comment by reaperhulk** [comment]
rebase time! :dancer:

**Comment by hynek** [comment]
Do we want to keep them in one env?  Is that to save time?

**Comment by lvh** [comment]
@hynek Yep, that's why.

**Comment by hynek** [comment]
You don’t have to pin the tools btw, if they break we wanna know.

**Comment by lvh** [comment]
I've merged forward and removed the irrelevant pinnings.

**Comment by lvh** [comment]
Please note that it does find some (alleged) problems:

```
~/P/pyopenssl: tox -e meta                                                                                                                                                                          10:24:52
GLOB sdist-make: /Users/lvh/Projects/pyopenssl/setup.py
meta inst-nodeps: /Users/lvh/Projects/pyopenssl/.tox/dist/pyOpenSSL-0.15.1.zip
meta runtests: PYTHONHASHSEED='1899853923'
meta runtests: commands[0] | pyflakes OpenSSL
OpenSSL/crypto.py:1534: local variable 'result_code' is assigned to but never used
OpenSSL/crypto.py:2598: '_ssl' imported but unused
OpenSSL/test/test_crypto.py:17: 'PY3' imported but unused
OpenSSL/test/test_crypto.py:18: redefinition of unused 'simplefilter' from line 9
OpenSSL/test/test_crypto.py:19: redefinition of unused 'catch_warnings' from line 9
OpenSSL/test/test_crypto.py:2286: local variable 'passwd' is assigned to but never used
OpenSSL/test/test_crypto.py:2287: local variable 'p12' is assigned to but never used
OpenSSL/test/test_ssl.py:2013: local variable 'conn' is assigned to but never used
OpenSSL/test/util.py:10: 'os' imported but unused
ERROR: InvocationError: '/Users/lvh/Projects/pyopenssl/.tox/meta/bin/pyflakes OpenSSL'
meta runtests: commands[1] | check-manifest
lists of files in version control and sdist match
meta runtests: commands[2] | pip-review
meta runtests: commands[3] | pyroma -d .
------------------------------
Checking .
Found pyOpenSSL
<string>:6: (WARNING/2) Definition list ends without a blank line; unexpected unindent.
------------------------------
Your package does not have keywords data.
You are using Setuptools or Distribute but do not specify if this package is zip_safe or not. You should specify it, as it defaults to True, which you probably do not want.
------------------------------
Final rating: 9/10
Cottage Cheese
------------------------------
_________________________________________________________________________________________________ summary __________________________________________________________________________________________________
ERROR:   meta: commands failed
```

**Comment by lvh** [comment]
thanks coveralls

File: tox.ini [hunk 1]
```diff
@@ -13,3 +13,16 @@ commands =
...
+     pyflakes
+     check-manifest
+     pip-tools
+     pyroma>=1.6 # <1.6 had bogus return values.
+commands =
+     pyflakes OpenSSL
+     check-manifest
+     pip-review
+     pyroma -d .
+
```
Comment (hynek) [inline tox.ini]: no empty line please here :)

**Comment by hynek** [comment]
Please add meta to travis and mark it as allowed to fail.  We’ll fix the problems separately. :)

File: tox.ini [hunk 2]
```diff
@@ -13,3 +13,16 @@ commands =
...
+[testenv:meta]
+deps =
+     pyflakes
+     check-manifest
+     pip-tools
+     pyroma>=1.6 # <1.6 had bogus return values.
+commands =
+     pyflakes OpenSSL
+     check-manifest
+     pip-review
```
Comment (hynek) [inline tox.ini]: I don’t think we need this since we don’t pin anything

**Comment by lvh** [comment]
Added the allowed failure on Travis. Ready for review again I think.

**Comment by hynek** [comment]
One more rebase please.  I’ve spent my free day fixing Travis. :-/

**Comment by lvh** [comment]
All done :)

**Comment by hynek** [comment]
So um…it seems like pyroma doesn’t get called at all if pyflakes fail so we don’t know what exactly happens once pyflakes pass. :)

I’m gonna say give em both separate tox targets.

**Comment by lvh** [comment]
done

**Comment by hynek** [comment]
You haven’t added pyroma to the matrix.  You’ve just allowed it to fail. :)

**Comment by lvh** [comment]
Oops, fixed & pushed. Thanks!

**Comment by hynek** [comment]
ok stupid question, will pyroma ever fail?  I mean it already has one bogus error message (version not being a string, probably because it’s read from a file).

**Comment by lvh** [comment]
It returns 0 on success, 1 on incorrect options and 2 if the rating is below 8. Our rating is 9, which is apparently Good Enough(TM). Alas, there doesn't appear to be an option to customize that...

**Comment by hynek** [comment]
So two thoughts:

1. that shouldn’t be marked as failok then, no?
2. how about using actually flake8 which is much more comprehesible and ignore certain errors like (E303 == too many free lines)?  My vim is currently borderline unusable because of all the errors so I would love to curb that a bit.

Cc @alex and @reaperhulk I guess.

(I would also love to change the empty line rule to PEP8 but baby steps)

**Comment by reaperhulk** [comment]
I like using `flake8` myself. We have an env named `pep8` on cryptography that runs flake8 (along with pep8-naming and @public's flake8-import-order that was built specifically for our project's import order preferences).

**Comment by lvh** [comment]
pyroma and flake8 check different things, no? pyroma is about packaging quality, flake8 is about code quality. They appear orthogonal.

**Comment by reaperhulk** [comment]
yeah flake8 is just pyflakes + pep8 + mccabe. pyroma is definitely orthogonal.

**Comment by hynek** [comment]
Yes, flake8 was intended instead of pyflakes, not pyroma.  Sorry for being ambiguous.

**Comment by hynek** [comment]
so to be clear: I think pyroma should not be in `allow_failures`.

**Comment by lvh** [comment]
OK. I will remove pyroma from allow_failures and use flake8 instead of pyflakes.

**Comment by hynek** [comment]
w00t!

Last step: we need to add

```ini
[flake8]
ignore = 303
```

to setup.cfg to ignore “too many blank lines” which are currently on purpose.

Any other suggestions that should be added?

Thanks Laurens for bearing with me. :)

**Comment by lvh** [comment]
done

**Comment by hynek** [comment]
long lines aren’t fine but that’s fortunately not what 303 means ;)

waiting for Travis now, thanks <3

**Comment by lvh** [comment]
:shipit:

**Comment by hynek** [comment]
Damit, it has to be `E303`, not `303`. :(  And I wondered why the errors still occurred! Sorryyy!

**Comment by lvh** [comment]
done

**Comment by hynek** [comment]
thanks, let’s finish this clusterfuck until I get some insane ideas :)

**Comment by lvh** [comment]
*revert merge button watching intensifies*

**Pull request merged**